feat: Add GitHub integration with agent_prompts and github_components

julian-risch opened the pull request:
### Related Issues

- Related to https://github.com/deepset-ai/haystack-experimental/issues/250
@sjrl brought up that one way to keep example files from the experimental Agent around is an integration https://github.com/deepset-ai/haystack-experimental/pull/263#issuecomment-2775755734

### Proposed Changes:

- Move github_components from experimental to a new integration
- Move agent_prompts from experimental to a new integration

The idea is to enable users to run the example notebook (or a version with updated imports) after having installed this new integration

### How did you test it?

Didn't test it for now. Just a draft.

### Notes for the reviewer

<!-- E.g. point out section where the reviewer  -->

### Checklist

- I have read the [contributors guidelines](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CODE_OF_CONDUCT.md)
- I have updated the related issue with new insights and changes
- I added unit tests and updated the docstrings
- I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.

sjrl commented on integrations/github/src/haystack_integrations/prompts/github/comment_tool.py:
```diff
@@ -0,0 +1,19 @@
+comment_prompt = """
```
More just a question, but for for variables like these I believe it's typical to make them in all caps, so COMMENT_PROMPT. What do you think?

julian-risch replied:
yes, make sense. 👍

julian-risch replied:
Done

sjrl commented on integrations/github/src/haystack_integrations/prompts/github/context.py:
```diff
@@ -0,0 +1,175 @@
+haystack_context_prompt = """
+
+Haystack-Agent was specifically designed to help developers with the Haystack-framework and any Haystack related
```
Remove extra new lines
```suggestion
haystack_context_prompt = """Haystack-Agent was specifically designed to help developers with the Haystack-framework and any Haystack related
```

julian-risch replied:
Done.

sjrl commented on integrations/github/src/haystack_integrations/prompts/github/file_editor_tool.py:
```diff
@@ -0,0 +1,120 @@
+file_editor_prompt = """
+Use the file editor to edit an existing file in the repository.
```
Fine to ignore this comment, but just so you are aware we are adding an extra new line at the beginning of the prompt by starting the text on the second line. 
```suggestion
file_editor_prompt = """Use the file editor to edit an existing file in the repository.
```

julian-risch replied:
Done.

sjrl commented on integrations/github/README.md:
```diff
@@ -0,0 +1,21 @@
+# github-haystack
```
I think it would be good for us to add examples here in the Readme on how to use or to link to the tutorial/google colab for how to use. 

Also another relevant detail I think is that these prompts were optimized using Anthropic models. Could be a useful thing for users to know.

julian-risch replied:
google colab in the cookbook and some more examples in an integration page is what I imagine. The README's we currently don't fill out, for example see: https://pypi.org/project/opensearch-haystack/
Might be a good idea to change that and use a copy of the integrations page. I don't see a good reason to keep it empty but I would prefer a consistent solution. I'll talk to Bilge.

sjrl commented on integrations/github/src/haystack_integrations/components/connectors/github/pr_creator.py:
```diff
@@ -0,0 +1,270 @@
+
+
+@component
+class GithubPRCreator:
```
I believe github is usually capitalized as GitHub, so worth considering updating our names I think
```suggestion
class GitHubPRCreator:
```

julian-risch replied:
Done.

sjrl commented:
@julian-risch maybe a general comment on the structure here. I see that the prompts aren't being used within the library and I understand they will be used in a future tutorial/colab. 

I wonder then if it would be helpful to instead pre-assemble the tools within the repo so users could easily import the tools and immediately pass them to an Agent. What do you think?

sjrl commented on integrations/github/src/haystack_integrations/components/connectors/github/file_editor.py:
```diff
@@ -0,0 +1,287 @@
+from base64 import b64decode, b64encode
```
We should also add the license header text to all python files right?

julian-risch replied:
Done.

sjrl commented on integrations/github/src/haystack_integrations/components/connectors/github/file_editor.py:
```diff
@@ -0,0 +1,287 @@
+    """
+
+    def __init__(
+        self,
+        github_token: Secret = Secret.from_env_var("GITHUB_TOKEN"),
```
Let me know what you think! If you agree, I'd advocated for updating the other components to do the same.

julian-risch replied:
I went for the first option and we now enforce kwargs in all components/tools of the GitHub integration.

sjrl commented on integrations/github/src/haystack_integrations/components/connectors/github/issue_viewer.py:
```diff
@@ -0,0 +1,212 @@
+            "User-Agent": "Haystack/GithubIssueViewer",
+        }
+
+    def _get_request_headers(self) -> dict:
+        """
+        Get headers with resolved token for the request.
+
+        :return: Dictionary of headers including authorization if token is present
+        """
```
I like this pattern for an incidental reason. By not putting the resolve of the Secret in the init method we allow users to not need to provide the secret until run time which is nice especially when running pipeline validation like in the deepset platform where the secrets are not available to the pipeline in the pipeline builder. 

Could be nice to do this in the other components, basically to move the resolve_value() outside of the init method.

julian-risch replied:
Done for all components.

sjrl commented:
@julian-risch overall this looks really good! I mostly have minor comments and only one larger conceptual one about maybe providing users Tools directly instead of needing to compose them, themselves. 

I didn't comb through every line since there is a lot, but it's well tested so it's good to go from my perspective! We can always make quick updates to this if things arise and depending on usage.

sjrl commented on integrations/github/src/haystack_integrations/tools/github/file_editor_tool.py:
```diff
@@ -0,0 +1,80 @@
+from haystack_integrations.prompts.github.file_editor_tool import FILE_EDITOR_PROMPT, FILE_EDITOR_SCHEMA
+
+
+class GitHubFileEditorTool(ComponentTool):
+    """
+    A Haystack tool for editing files in GitHub repositories.
+    """
```
Oh interesting I wasn't thinking to inherit from ComponentTool but do something like 
```python
GitHubFileEditorTool = ComponentTool(...)
```
and then people could import the pre-made `GitHubFileEditorTool` but I can see how this version would be more customizable.

sjrl commented on integrations/github/src/haystack_integrations/tools/github/file_editor_tool.py:
```diff
@@ -0,0 +1,80 @@
+            parameters=parameters,
+        )
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Serializes the tool to a dictionary.
+
+        :returns:
+            Dictionary with serialized data.
+        """
+        return default_to_dict(
+            self,
+            name=self.name,
+            description=self.description,
+            parameters=self.parameters,
+            github_token=self.github_token.to_dict(),
+            repo=self.repo,
+            branch=self.branch,
+            raise_on_failure=self.raise_on_failure,
+        )
```
If we go this route of inheriting from ComponentTool, we won't be able to use this `to_dict` method I think. At least we use different sede methods for Tools with this dict structure 
```python
{"type": generate_qualified_class_name(type(self)), "data": serialized}
```
so we'd probably need to follow that as well right? 

Since when deserializing this in a pipeline I believe we will eventually call `deserialize_tools_or_toolset_inplace`. Do these methods work in that case?

sjrl replied:
@julian-risch okay interesting that this works since you added the pipeline serialization test.

I still wonder to be consistent if we should not rename `init_parameters` to `data` in the serialized dict since that appears to be the pattern we use in our other tools and expect in `deserialize_tools_or_toolset_inplace`

julian-risch commented:
@sjrl I added a test called `test_pipeline_serialization`, added `_get_request_headers` to all components and here is the example notebook with updated code up until the GitHub token is required. I commented out `"message": {"source": "documents", "handler": message_handler},` because it didn't work for me and need to ask @mathislucka for advice.

https://colab.research.google.com/drive/1ktlwQ-CDLGDs2uZXvzgG8XspfjPidYqZ?usp=sharing

@sjrl If GitHubFileEditorTool looks good to you, I will add tools for all other components and probably update the directory structure a bit.

sjrl commented on integrations/github/src/haystack_integrations/tools/github/issue_commenter_tool.py:
```diff
@@ -1,19 +1,46 @@
     """
     A tool for commenting on GitHub issues.
+
+    :param name: Optional name for the tool.
+    :param description: Optional description.
```
I do realize this is a bit confusing in our tools, but it seems that if we define a `__init__` then these docstrings are put under the `__init__` def. If there is no `__init__` defined like in `Tool` then we put it in the class description.

julian-risch replied:
Ah, of course! Do you think we should a usage example here then (in addition to moving the param docstrings to the init)? I realized that's missing too.

sjrl commented on .github/workflows/github.yml:
```diff
@@ -0,0 +1,75 @@
+        if: matrix.python-version == '3.9' && runner.os == 'Linux'
+        run: hatch run docs
+
+      - name: Run tests
+        run: hatch run cov-retry
```
Oh it could also be good to set up the new testing Stefano and David have been adding to the other integrations. Here is an example https://github.com/deepset-ai/haystack-core-integrations/pull/1828
